### ISSUE TITLE
loss function different from the paper

### ISSUE THREAD

**Issue by dawnranger** [snowkylin/line#2]
Thanks for your elegant implementation of LINE. 

I notice that  loss function  in your code:

> self.inner_product = tf.reduce_sum(self.u_i_embedding * self.u_j_embedding, axis=1)
self.loss = -tf.reduce_mean(tf.log_sigmoid(self.label * self.inner_product))

seems to implement a loss function like this:

<a href="https://www.codecogs.com/eqnedit.php?latex=\log\sigma(u_j^Tu_i&space;-&space;\sum_{n=1}^Ku_n^Tu_i)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\log\sigma(u_j^Tu_i&space;-&space;\sum_{n=1}^Ku_n^Tu_i)" title="\log\sigma(u_j^Tu_i - \sum_{n=1}^Ku_n^Tu_i)" /></a>

which is  different from the function in your slide:

<a href="https://www.codecogs.com/eqnedit.php?latex=\log\sigma(u_j^Tu_i)&space;&plus;&space;\sum_{n=1}^K\log\sigma(-u_n^Tu_i)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\log\sigma(u_j^Tu_i)&space;&plus;&space;\sum_{n=1}^K\log\sigma(-u_n^Tu_i)" title="\log\sigma(u_j^Tu_i) + \sum_{n=1}^K\log\sigma(-u_n^Tu_i)" /></a>

However, the embedding learned by this code is feasible in my experiment, could anybody explain this? Thanks in advance.

**Comment by snowkylin** [comment]
Please notice that after calculating `tf.log_sigmoid(self.label * self.inner_product)` in TensorFlow, the shape of the tensor is still [num_batch], just the same as `self.inner_product`. That is, all elements in `self.inner_product` are processed by `tf.log_sigmoid()` operation and you will get a vector of `\log \sigma (u_j^T u_i)`. The `tf.reduce_mean()` (maybe you can replace it with `tf.reduce_sum()`) operation is the last operation before getting the loss, just the same as the equation which is shown in the paper and my slide.

**Comment by dawnranger** [comment]
You are absolutely right. I misunderstand your code. Thank you!
